include/vinecopulib/bicop/implementation/extreme_value.ipp

```diff
@@ -0,0 +1,111 @@
+// Copyright © 2016-2023 Thomas Nagler and Thibault Vatter
+//
+// This file is part of the vinecopulib library and licensed under the terms of
+// the MIT license. For a copy, see the LICENSE file in the root directory of
+// vinecopulib or https://vinecopulib.github.io/vinecopulib/.
+
+#include <vinecopulib/misc/tools_eigen.hpp>
+#include <vinecopulib/misc/tools_integration.hpp>
+
+namespace vinecopulib {
+inline Eigen::VectorXd ExtremeValueBicop::cdf(
+    const Eigen::MatrixXd &u
+)
+{
+    auto f = [this](const double &u1, const double &u2) {
+        double t = std::log(u2) / std::log(u1 * u2);
+        t = pickands(t);
+        t = (std::log(u1) + std::log(u2)) * t;
+        return std::exp(t);
+    };
+    return tools_eigen::binaryExpr_or_nan(u, f);
```

Comment:
I know we've been doing this everywhere. But it feels so wrong in retrospective. It must be terribly inefficient when compared to truly vectorized operations no?

Reply:
probably yes. we should benchmark at some point and change everywhere if there's a big difference. I think I tested in a very early stage and saw no difference, but my benchmarking was not exactly scientific.

Reply:
I used the following to test unary ops:

```
#include <iostream>
#include <Eigen/Dense>
#include <chrono>
#include <cmath>
#include <vector>
#include <numeric>

using namespace Eigen;
using namespace std;
using namespace std::chrono;

// Function to compute the mean of a vector
double mean(const vector<double> &v)
{
  return accumulate(v.begin(), v.end(), 0.0) / v.size();
}

// Function to compute the standard error of a vector
double standard_error(const vector<double> &v)
{
  double m = mean(v);
  double sq_sum = inner_product(v.begin(), v.end(), v.begin(), 0.0, [](double a, double b)
                                { return a + b; }, [m](double a, double b)
                                { return (a - m) * (b - m); });
  return sqrt(sq_sum / v.size()) / sqrt(v.size());
}

// Template function to benchmark a coefficient-wise operation
template <typename F>
void benchmark_coeff_op(MatrixXd &A, MatrixXd &B, F func, const string &func_name)
{
  vector<double> durations;
  for (int i = 0; i < 5; ++i)
  {
    auto start = high_resolution_clock::now();
    B = A.array().unaryExpr(func);
    auto end = high_resolution_clock::now();
    auto duration = duration_cast<milliseconds>(end - start).count();
    durations.push_back(duration);
  }
  cout << "Coefficient-wise operation (" << func_name << ") took: "
       << mean(durations) << " ms ± " << standard_error(durations) << " ms" << endl;
}

// Template function to benchmark a unary operation
template <typename F>
void benchmark_unary_op(MatrixXd &A, MatrixXd &B, F func, const string &func_name)
{
  vector<double> durations;
  for (int i = 0; i < 5; ++i)
  {
    auto start = high_resolution_clock::now();
    B = A.unaryExpr(func);
    auto end = high_resolution_clock::now();
    auto duration = duration_cast<milliseconds>(end - start).count();
    durations.push_back(duration);
  }
  cout << "Unary operation (" << func_name << ") took: "
       << mean(durations) << " ms ± " << standard_error(durations) << " ms" << endl;
}

int main()
{
  const int size = 10000;
  MatrixXd A = MatrixXd::Random(size, size);
  MatrixXd B(size, size);

  // Benchmark for different math functions
  benchmark_coeff_op(A, B, [](double x)
                     { return sin(x); }, "sin");
  benchmark_unary_op(A, B, [](double x)
                     { return sin(x); }, "sin");

  benchmark_coeff_op(A, B, [](double x)
                     { return cos(x); }, "cos");
  benchmark_unary_op(A, B, [](double x)
                     { return cos(x); }, "cos");

  benchmark_coeff_op(A, B, [](double x)
                     { return exp(x); }, "exp");
  benchmark_unary_op(A, B, [](double x)
                     { return exp(x); }, "exp");

  benchmark_coeff_op(A, B, [](double x)
                     { return log(x + 1.0); }, "log(x + 1)"); // log(x + 1) to avoid log(0)
  benchmark_unary_op(A, B, [](double x)
                     { return log(x + 1.0); }, "log(x + 1)"); // log(x + 1) to avoid log(0)

  return 0;
}
```

And the following for binary ops:

```
#include <iostream>
#include <Eigen/Dense>
#include <chrono>
#include <cmath>
#include <vector>
#include <numeric>

using namespace Eigen;
using namespace std;
using namespace std::chrono;

// Function to compute the mean of a vector
double mean(const vector<double> &v)
{
  return accumulate(v.begin(), v.end(), 0.0) / v.size();
}

// Function to compute the standard error of a vector
double standard_error(const vector<double> &v)
{
  double m = mean(v);
  double sq_sum = inner_product(v.begin(), v.end(), v.begin(), 0.0, [](double a, double b)
                                { return a + b; }, [m](double a, double b)
                                { return (a - m) * (b - m); });
  return sqrt(sq_sum / v.size()) / sqrt(v.size());
}

// Template function to benchmark arithmetic operations
template <typename F>
void benchmark_arith_op(int size, F func, const string &func_name)
{
  vector<double> durations;
  for (int i = 0; i < 5; ++i)
  {
    // Set seed for reproducibility
    srand(i);
    MatrixXd A = MatrixXd::Random(size, size);
    MatrixXd B = MatrixXd::Random(size, size);
    MatrixXd C(size, size);

    auto start = high_resolution_clock::now();
    C = func(A, B);
    auto end = high_resolution_clock::now();
    auto duration = duration_cast<milliseconds>(end - start).count();
    durations.push_back(duration);
  }
  cout << "Arithmetic operation (" << func_name << ") took: "
       << mean(durations) << " ms ± " << standard_error(durations) << " ms" << endl;
}

// Template function to benchmark binaryExpr operations
template <typename F>
void benchmark_binary_expr(int size, F func, const string &func_name)
{
  vector<double> durations;
  for (int i = 0; i < 5; ++i)
  {
    // Set seed for reproducibility
    srand(i);
    MatrixXd A = MatrixXd::Random(size, size);
    MatrixXd B = MatrixXd::Random(size, size);
    MatrixXd C(size, size);

    auto start = high_resolution_clock::now();
    C = A.binaryExpr(B, func);
    auto end = high_resolution_clock::now();
    auto duration = duration_cast<milliseconds>(end - start).count();
    durations.push_back(duration);
  }
  cout << "Binary operation (" << func_name << " with binaryExpr) took: "
       << mean(durations) << " ms ± " << standard_error(durations) << " ms" << endl;
}

int main()
{
  const int size = 10000;

  // Benchmark for addition
  benchmark_arith_op(size, [](const MatrixXd &X, const MatrixXd &Y)
                     { return X + Y; }, "addition");
  benchmark_binary_expr(size, [](double x, double y)
                        { return x + y; }, "addition");

  // Benchmark for subtraction
  benchmark_arith_op(size, [](const MatrixXd &X, const MatrixXd &Y)
                     { return X - Y; }, "subtraction");
  benchmark_binary_expr(size, [](double x, double y)
                        { return x - y; }, "subtraction");

  // Benchmark for multiplication
  benchmark_arith_op(size, [](const MatrixXd &X, const MatrixXd &Y)
                     { return X.array() * Y.array(); }, "multiplication");
  benchmark_binary_expr(size, [](double x, double y)
                        { return x * y; }, "multiplication");

  // Benchmark for division
  benchmark_arith_op(size, [](const MatrixXd &X, const MatrixXd &Y)
                     { return X.array() / Y.array(); }, "division");
  benchmark_binary_expr(size, [](double x, double y)
                        { return x / y; }, "division");

  return 0;
}
```

........ and it seems that the difference ain't big indeed. But I might be doing something wrong.